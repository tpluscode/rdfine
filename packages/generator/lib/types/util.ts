import { GraphPointer } from 'clownface'
import { rdf, rdfs, schema } from '@tpluscode/rdf-ns-builders'
import { TypeMap } from '../types/index.js'

export function isDatatype(term: GraphPointer, typeMappings: TypeMap): boolean {
  if (term.has(rdf.type, rdfs.Datatype).values.length) {
    return true
  }

  const result = term.has(rdf.type, [schema.DataType, rdfs.Datatype]).values.length > 0
  if (result) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isDatatype(superClass, typeMappings)) {
      return true
    }
  }

  return false
}

export function isEnumerationType(term: GraphPointer) {
  if (term.has(rdfs.subClassOf, schema.Enumeration).values.length > 0) {
    return true
  }

  const superClasses = term.out(rdfs.subClassOf).toArray()
  for (const superClass of superClasses) {
    if (isEnumerationType(superClass)) {
      return true
    }
  }

  return false
}
