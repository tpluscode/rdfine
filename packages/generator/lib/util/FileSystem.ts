import path from 'path'
import { RealFileSystemHost } from '@ts-morph/common'

export default class FileSystem extends RealFileSystemHost {
  private readonly outDir: string

  constructor(outDir: string) {
    super()
    this.outDir = outDir
  }

  getCurrentDirectory(): string {
    return path.resolve(process.cwd(), this.outDir)
  }
}
