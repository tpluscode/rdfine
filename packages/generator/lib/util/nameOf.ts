import { shrink } from '@zazuko/rdf-vocabularies'
import { GraphPointer } from 'clownface'

const rdfineBaseNames = ['equals', 'types', 'pointer', 'unionGraphPointer', '_graphId', '_parent', 'isAnonymous', 'hasType', 'get', 'getArray', 'getBoolean', 'getString', 'getNumber', 'toJSON', '_getObjects', '_create']

export function nameOf(term: GraphPointer, safeIdentifier = false): string {
  const shrunk = shrink(term.value)

  const name = shrunk.split(':')[1]
  if (rdfineBaseNames.includes(name) && safeIdentifier) {
    return `_${name}`
  }
  return name
}
