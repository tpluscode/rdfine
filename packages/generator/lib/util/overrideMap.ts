import { expand } from '@zazuko/rdf-vocabularies'

export function expandMapKeys<T>(overrideMap: Record<string, T>, namespace: string): Record<string, T> {
  return Object.entries(overrideMap)
    .reduce((newMap, [key, override]) => {
      let typeUri: string
      if (key.includes(':')) {
        typeUri = expand(key)
      } else {
        typeUri = `${namespace}${key}`
      }

      return {
        ...newMap,
        [typeUri]: override,
      }
    }, {})
}
