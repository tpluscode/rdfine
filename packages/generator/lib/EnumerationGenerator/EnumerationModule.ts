import { ObjectLiteralExpression, Project } from 'ts-morph'
import { SingleContextClownface } from 'clownface'
import { rdf } from '@tpluscode/rdf-ns-builders'
import { Context, GeneratedModule } from '../index'
import { EnumerationType, TypeMetaCollection } from '../types'

export class EnumerationModule implements GeneratedModule {
  type: EnumerationType
  node: SingleContextClownface;

  public constructor(enumeration: SingleContextClownface, type: EnumerationType) {
    this.node = enumeration
    this.type = type
  }

  writeModule(project: Project, types: TypeMetaCollection, context: Pick<Context, 'prefix' | 'log'>) {
    context.log.debug('Generating enumeration %s', this.type.name)
    const enumFile = project.createSourceFile(`${this.type.module}.ts`, {}, { overwrite: true })

    enumFile.addImportDeclaration({
      namedImports: ['NamedNode'],
      moduleSpecifier: 'rdf-js',
    })
    enumFile.addImportDeclaration({
      namedImports: [context.prefix],
      moduleSpecifier: './lib/namespace',
    })

    const enumeration = enumFile.addExportAssignment({
      expression: '{}',
      isExportEquals: false,
    }).getExpression() as ObjectLiteralExpression

    let enumerationType = ''
    this.node.in(rdf.type)
      .forEach(enumMember => {
        const memberType = types.get(enumMember)
        if (memberType?.type !== 'EnumerationMember') {
          context.log.warn('Term %s was not recognized as enumeration member', enumMember.value)
          return
        }

        context.log.debug('Adding enum member %s', memberType.prefixedName)

        const nodeType = `NamedNode<'${enumMember.value}'>`
        enumeration.addPropertyAssignment({
          name: memberType.termName,
          initializer: memberType.prefixedName,
          trailingTrivia: ` as ${nodeType}`,
        })

        if (enumerationType === '') {
          enumerationType = `\n${nodeType}`
        } else {
          enumerationType += `\n| ${nodeType}`
        }
      })

    enumFile.addTypeAlias({
      name: this.type.name,
      type: enumerationType || 'NamedNode',
      isExported: true,
    })

    return {
      mainModuleExport: this.type.name,
    }
  }
}
