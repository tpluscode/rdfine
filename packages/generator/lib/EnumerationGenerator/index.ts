import { rdf, rdfs } from '@tpluscode/rdf-ns-builders'
import { Context, GeneratedModule } from '../index.js'
import { EnumerationModule } from './EnumerationModule.js'
import { TypeMetaCollection } from '../types/index.js'

export function findTermsToGenerate(types: TypeMetaCollection, { vocabulary }: Pick<Context, 'vocabulary'>): GeneratedModule[] {
  return vocabulary
    .has(rdf.type, rdfs.Class)
    .filter(term => types.get(term)?.type === 'Enumeration')
    .map(enumeration => {
      const meta = types.getOrThrow(enumeration)
      if (meta.type !== 'Enumeration') {
        throw new Error(`Expected to find an Enumeration type by got '${meta.type}'`)
      }

      return new EnumerationModule(enumeration, meta)
    })
}
