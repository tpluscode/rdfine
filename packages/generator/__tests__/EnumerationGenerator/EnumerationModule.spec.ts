import cf, { Clownface } from 'clownface'
import $rdf from 'rdf-ext'
import { rdf } from '@tpluscode/rdf-ns-builders'
import { Project } from 'ts-morph'
import { EnumerationModule } from '../../lib/EnumerationGenerator/EnumerationModule'
import { fakeLog } from '../_helpers/util'
import { FakeTypeCollection } from '../_helpers/FakeTypeCollection'
import { ex } from '../_helpers/prefix'

describe('EnumerationModule', () => {
  let project: Project
  let graph: Clownface

  beforeEach(() => {
    project = new Project({
      useInMemoryFileSystem: true,
    })

    graph = cf({ dataset: $rdf.dataset() })
  })

  it('writes a module with enumeration members', () => {
    // given
    const types = new FakeTypeCollection([
      [ex.Member1, { type: 'EnumerationMember', termName: 'Member1', prefixedName: 'ex.Member1' }],
    ])

    const enumeration = graph.node(ex.Enum)
      .addIn(rdf.type, [ex.Member1, ex.Member2])
    const module = new EnumerationModule(enumeration, {
      name: 'Enum',
      module: './Enum',
      qualifiedName: 'Ex.Enum',
      type: 'Enumeration',
    })

    // when
    module.writeModule(project, types, {
      log: fakeLog(),
      prefix: 'ex',
    })

    // then
    expect(project.getSourceFile('./Enum.ts')).toMatchSnapshot()
  })

  it('uses plain NamedNode for enumerations without members', () => {
    // given
    const types = new FakeTypeCollection([
    ])

    const enumeration = graph.node(ex.Enum)
      .addIn(rdf.type, [ex.Member1, ex.Member2])
    const module = new EnumerationModule(enumeration, {
      name: 'Enum',
      module: './Enum',
      qualifiedName: 'Ex.Enum',
      type: 'Enumeration',
    })

    // when
    module.writeModule(project, types, {
      log: fakeLog(),
      prefix: 'ex',
    })

    // then
    expect(project.getSourceFile('./Enum.ts')).toMatchSnapshot()
  })
})
