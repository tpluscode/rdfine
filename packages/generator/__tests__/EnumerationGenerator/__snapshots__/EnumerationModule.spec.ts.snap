// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnumerationModule uses plain NamedNode for enumerations without members: toMatchSnapshot 1`] = `
"import { NamedNode } from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";

export default {};

export type Enum = NamedNode;
"
`;

exports[`EnumerationModule writes a module with enumeration members: toMatchSnapshot 1`] = `
"import { NamedNode } from \\"rdf-js\\";
import { ex } from \\"./lib/namespace\\";

export default {
    Member1: ex.Member1 as NamedNode<'http://example.com/Member1'>,
    Member2: ex.Member2 as NamedNode<'http://example.com/Member2'>
};

export type Enum = 
    NamedNode<'http://example.com/Member1'>
    | NamedNode<'http://example.com/Member2'>;
"
`;
